lib/srv/regular/sshserver.go

```diff
@@ -297,6 +328,28 @@ func (s *Server) Shutdown(ctx context.Context) error {
 	return err
 }
 
+func (s *Server) acquireSemaphoreStartUserCleanup() {
+	s.startCleanupOnce.Do(func() {
+		semLock, err := services.AcquireSemaphoreLock(s.ctx,
```

Comment:
Hmm, I'm not sure I understand how acquiring the lock here helps?

As currently implemented (if I'm reading correctly), it will just try to acquire the lock for 10 minutes when SSH server starts (i.e. when Teleport process starts), then will run the same UserCleanup loop as before. So it will either just start UserCleanup as usual, or be a no-op if the process restarted within 10 minutes and not start it at all.

How does this help prevent race conditions between user creation and deletion? I thought that we would need to synchronize initialization of new SSH session and UserCleanup so either one or the other runs at the same time.

Am I missing something? cc @nklaassen @espadolini

Reply:
@lxea @espadolini @nklaassen I'm also thinking (perhaps a bit late thought) that since we only need to synchronize it within the process, there's no real need to do it through the backend? Semaphore was primarily implemented for distributed locks. Surely we can just use an in-memory lock/mutex? I know I suggested semaphore myself before and it's fine if it's easier, but I'd see if using a simple in-memory lock would be simpler.

Still I'm not sure that we're currently acquiring semaphore in the right place as mentioned above though.

Reply:
We currently run multiple instances of the `TeleportProcess` either during CA rotations or during SIGHUP upgrades, so I'd very much like the lock to be stored in the process storage backend (which would be shared between all of them).

Reply:
A session tracker wouldn't quite do it across teleport processes - we don't have (I think) a way to fetch the sessions on a specific node, so checking the session trackers would be expensive.

Maybe a semaphore for each created unix user? This feels more complicated than just parking a process with the correct UID before launching the intended process, however.